Implement Partitioned on Session Cookie

volosied opened the pull request:
fixes #27798

Beta guard & cookiePartitioned metatype is marked internal for now. 

Some notes: 

1)

> The default value for cookiePartitioned (in the code) is null which means a user has not defined a value.
> 
> This allows a user to partition all cookies except for the session cookie. (see `testHttpSessionOverridesHttpChannelConfig`).  If the default value was false, then we wouldn't be able to differentiate the default false value from a user explict false value. 
> 
> This change also modified the logic for when partitioned is set.  If partitioned is set to false on the HttpCookie object or is set to null, it will not be generated (see CookieUtils).  Any other value will add the partitioned value to the cookie (i.e empty string, "ABC", etc) 

2)

> WebContainerRequestState was modified to support multiple attributes on a given cookie which is why cookieAttributesMap is now a  type: `Map<String,HashMap<String,String>>`.  The getCookieAttributes has been modified to return colon separated values (i.e `attributeName1=attributeValue1;attributeName2=attributeValue2`).

volosied committed:
Implement Partitioned on Session Cookie

Update WebContainerRequestState#setCookieAttributes to support Partitioned Attribute

volosied commented on dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java:
```diff
@@ -216,10 +214,16 @@ public void setCurrentThreadsIExtendedResponse(IExtendedResponse hres) {
+     * must exist.
+     * 
+     * If the cookieAttribute is a flag, then it should in the form: attributeName=true
+     * If the attributeValue evaluates to true via Boolean.parseBoolean() then 
+     * only the cookieAttribute will be sent. For example, "The call setCookieAttributes("myCookie", "secure=true") 
```
I don't think this is necessary.  Maybe say "If setting partitioned, then use true as the value. i.e "partitioned=true".

volosied commented on dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java:
```diff
@@ -231,40 +235,56 @@ public void setCookieAttributes(String cookieName, String cookieAttributes) {
                 return;
         }
+
+        // Should we throw any exceptions? such as: 
+        // if(attribute.length == 1) { // No attributeValue set
+        //     throw IllegalStateException("attribute value not specififed");
+        // }
```
This would require a API change since setCookieAttributes doesn't throw any exceptions.

volosied commented on dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java:
```diff
@@ -231,40 +235,56 @@ public void setCookieAttributes(String cookieName, String cookieAttributes) {
         }
 
-        String cookieAttributes = cookieAttributesMap.get(cookieName);
+        String cookieAttributes = cookieAttributesMap.get(cookieName).entrySet()
+                                                     .stream().map(e->e.getKey() + "=" + e.getValue())
+                                                     .collect(java.util.stream.Collectors.joining(";"));
```
We're using Java 8 as a min, so this I opted to use stream as it's a lot easier make the return value.

volosied commented on dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java:
```diff
@@ -40,7 +37,8 @@ public class WebContainerRequestState {
     private IExtendedResponse currentThreadsIExtendedResponse;
-    private Map<String,String> cookieAttributesMap = null;
+    // Each cookie has a Map of attributes
+    private Map<String,HashMap<String,String>> cookieAttributesMap = null;
```
Nested map should be just be `Map`, not `HashMap`

volosied replied:
Created compilation errors, so I'll leave it as it is: `Map<String,HashMap<String,String>>`

zhang-ca reviewed (dismissed):
New session property cookiePartitioned looks fine to me. Thanks

volosied committed:
Beta check and mark session config cookiePartitioned as internal

volosied committed:
Create Session Tests

Add paritioned to HttpSession config in fattest.simplicity

Add testHttpSessionOverridesHttpChannelConfig

volosied committed:
HttpSession Partitioned: Override HttpChannel Config

volosied committed:
Clean up / Address review comments

Apply sugguestion: set partitioned to false for all scenarios except SameSite=None

Reword cookie.partitioned.desc

volosied commented:
#build

volosied commented:
#build #spawn.fullfat.buckets=com.ibm.ws.webcontainer.servlet.4.0_fat

pnicolucci commented on dev/com.ibm.ws.webcontainer.servlet.4.0_fat/publish/servers/servlet40_partitioned/server.xml:
```diff
@@ -8,7 +8,7 @@
  -->
 <server description="Partitioned Cookies Test">
-
+    <!-- No SameSite Configurations in this server.xml becuase they are all setup individually in each test -->
```
`becuase` -> `because`

zhang-ca commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/session/impl/SessionAffinityManagerImpl.java:
```diff
@@ -510,6 +508,17 @@ public Cookie run() {
+                                if(sessionSameSiteCookie.equals(SameSiteCookie.NONE) && _smc.getSessionCookiePartitioned()) {
+                                    requestState.setCookieAttributes(cookie.getName(), "Partitioned=true");
+                                } else if(!_smc.getSessionCookiePartitioned()) { // if set to false
+                                    requestState.setCookieAttributes(cookie.getName(), "Partitioned=false");
```
If not SameSiteCookie.NONE, Partitioned=false is allowed? 
I thought only applies when samesite none is set.

zhang-ca replied:
```
                    if(_smc.getSessionCookiePartitioned() != null && sessionSameSiteCookie.equals(SameSiteCookie.NONE)) {
                        if(_smc.getSessionCookiePartitioned()) {
                            requestState.setCookieAttributes(cookie.getName(), "Partitioned=true");
                        } else { // if set to false
                            requestState.setCookieAttributes(cookie.getName(), "Partitioned=false");
                        }
                    }
```

volosied commented:
#build

volosied commented:
#build

pnicolucci commented on dev/com.ibm.ws.session/src/com/ibm/ws/session/SessionProperties.java:
```diff
@@ -760,6 +759,14 @@ private static boolean setDefaultConfiguration(SessionManagerConfig smc, Map<Str
+            s = "cookiePartitioned";
+            bValue = propertyToBoolean(xtpProperties.get(s));
+            if (bValue != null) {
+                smc.setSessionCookiePartitioned(bValue.booleanValue());
```
Why call booleanValue here when setSessionCookieParitioned takes a `Boolean`?

volosied replied:
Good point -- I'll make the fix

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:
```diff
@@ -2924,10 +2924,12 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
-                            Tr.debug(tc, "[1] Setting the Partitioned attribute for SameSite=None");
-                        }
-                        cookie.setAttribute("partitioned", "");
+                            if(cookie.getAttribute("partitioned") == null) {  // null means no value has been set yet
```
Any need for beta checks here?

volosied replied:
This check is nested within the condition: "getServiceContext().getHttpConfig().getPartitioned() == true" (httpEndpoint samesite config) which is internal and beta guarded. 

I can add a check if you prefer?

pnicolucci replied:
if we're eventually guarded then I'm fine with it as is.

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/HttpBaseMessageImpl.java:
```diff
@@ -2941,10 +2943,12 @@ else if (getServiceContext().getHttpConfig().onlySameSiteStar()) {
+                    if(cookie.getAttribute("partitioned") == null) {  // null means no value has been set yet
+                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                            Tr.debug(tc, "[2] Setting the Partitioned attribute for SameSite=None");
+                        }
```
Any need for beta checks in this area of code?

volosied replied:
same reasoning as in my other comment above -- getHttpConfig().getPartitioned() is beta guarded , so I don't think it needs it here.

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/CookieUtils.java:
```diff
@@ -393,7 +390,7 @@ private static String convertV0SetCookie(HttpCookie cookie, boolean isRFC1123Dat
         if(ProductInfo.getBetaEdition()) {
             value = cookie.getAttribute("partitioned");
-            if (null != value) {
+            if (null != value && !value.equalsIgnoreCase("false")) {
```
Why not just check for value.equalsIgnoreCase("true")?

volosied replied:
I set it to an empty thing is some cases, but I guess it could be set to "true" there as well. 
https://github.com/OpenLiberty/open-liberty/pull/27843/files#diff-d0a69db5d3911d95a2ca377b482ad2c1d9463356ed2a2a4d5e834e54151df4edR2950


My reason is that since this is just a attribute and doesn't need a value, the value can be anything. ("true", "", "dummyValue", etc). However, false is explicit in that you don't want it set.

volosied replied:
Customers could use "" or "true" if they're on servlet 6.

pnicolucci replied:
ok that's fair

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/CookieUtils.java:
```diff
@@ -476,7 +473,7 @@ private static String convertV1SetCookie(HttpCookie cookie) {
         if(ProductInfo.getBetaEdition()) {
             value = cookie.getAttribute("partitioned");
-            if (null != value) {
+            if (null != value && !value.equalsIgnoreCase("false")) {
```
Why not just check for true?

pnicolucci commented on dev/com.ibm.ws.transport.http/src/com/ibm/ws/http/channel/internal/cookies/CookieUtils.java:
```diff
@@ -575,7 +572,7 @@ private static String convertV1SetCookie2(HttpCookie cookie) {
         if(ProductInfo.getBetaEdition()) {
             value = cookie.getAttribute("partitioned");
-            if (null != value) {
+            if (null != value && !value.equalsIgnoreCase("false")) {
```
Why not just check for true?

pnicolucci commented on dev/com.ibm.ws.webcontainer/src/com/ibm/wsspi/webcontainer/WebContainerRequestState.java:
```diff
@@ -231,19 +230,35 @@ public void setCookieAttributes(String cookieName, String cookieAttributes) {
-        if (!attribute[0].equals("SameSite")) {
+        /*
+         * Beta: This method will start accepting "Parititioned" as a new attribute. It will need to be in the form "partitioned=boolean".
+         */
```
It needs to be `partitioned=true/false` is a bit more clear IMO

volosied replied:
I can update it

volosied commented on dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/session/impl/SessionAffinityManagerImpl.java:
```diff
@@ -510,6 +508,17 @@ public Cookie run() {
                         String sameSiteCookieValue = sessionSameSiteCookie.getSameSiteCookieValue();
                         requestState.setCookieAttributes(cookie.getName(), "SameSite=" + sameSiteCookieValue);
+
+                        if (ProductInfo.getBetaEdition()) {
+                            // not null means a user defined config was set
+                            if(_smc.getSessionCookiePartitioned() != null) {
+                                if(sessionSameSiteCookie.equals(SameSiteCookie.NONE) && _smc.getSessionCookiePartitioned()) {
+                                    requestState.setCookieAttributes(cookie.getName(), "Partitioned=true");
+                                } else if(!_smc.getSessionCookiePartitioned()) { // if set to false
+                                    requestState.setCookieAttributes(cookie.getName(), "Partitioned=false");
+                                }
+                            }
+                        }
```
Log a message here

zhang-ca reviewed (dismissed):
Looks good from session manager perspective.

pnicolucci commented:
@volosied  one last thing, is there anything we need to address for : https://jakarta.ee/specifications/servlet/6.0/apidocs/jakarta.servlet/jakarta/servlet/sessioncookieconfig / the web.xml configuration for the session cookie?

volosied commented:
#build

pnicolucci commented:
#libby

jantley-ibm commented on dev/com.ibm.ws.session/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -61,6 +58,10 @@ cookie.same.site.lax=Set the session cookie SameSite attribute value to Lax
 
+# The word Partitioned should not be translated
+cookie.partitioned=Include a Partitioned attribute in session cookies
+cookie.partitioned.desc=Specifies that session cookies include the partitioned attribute. Only applies when samesite none is set. The partitioned attribute will only be used by browsers which support cookie partitioning.
```
Suggested rewrite:
cookie.partitioned.desc=Specifies that session cookies include the Partitioned attribute. Only applies when the SameSite attribute is set to none. The Partitioned attribute is only used by browsers that support cookie partitioning.

jantley-ibm replied:
Correction:
cookie.partitioned.desc=Specifies that session cookies include the Partitioned attribute. The Partitioned attribute only applies when the SameSite attribute is set to none, and is only used by browsers that support cookie partitioning.

volosied replied:
Thanks -- I've made the update

jantley-ibm reviewed (commented):
Suggested rewrite included. Let me know if you have any questions about the proposed changes. I will set the ID reviewed tag when we agree on changes. Thanks --

volosied commented:
#libby
